crates/smelt-graph/src/executor/profiler.rs

```diff
@@ -59,6 +64,70 @@ fn sample_memory_and_load(ppid: u32) -> Option<SampleStruct> {
     })
 }
 
+
+async fn docker_sample(
+    docker_client: &Docker,
+    command_ref: &String,
+) -> Option<SampleStruct> {
+    let stats = docker_client.stats(&command_ref, Some(StatsOptions {
+        stream: false,
+        ..Default::default()
+    })).try_collect::<Vec<_>>().await.unwrap();
+
+    for stat in stats {
+        return Some(SampleStruct {
+            memory_used: stat.memory_stats.usage.unwrap_or(0),
+            cpu_time_delta: stat.cpu_stats.cpu_usage.total_usage - stat.precpu_stats.cpu_usage.total_usage
+        });
+    }
+    None
+}
+
+
+pub async fn profile_cmd_docker(
+    tx: Sender<Event>,
+    docker_client: Docker,
+    sample_freq_ms: u64,
+    command_ref: String,
+    trace_id: String,
+) {
+    let mut prev_sample = None;
+    let mut prev_sample_time = Instant::now();
+
+    loop {
+        let new_sample_time = Instant::now();
+        let new_sample = docker_sample(&docker_client, &command_ref).await;
+        let sampling_duration = u64::try_from(
+            Instant::now().duration_since(new_sample_time).as_millis()).unwrap_or(0);
+
+        if let Some(ref sample) = new_sample {
+            if let Some(ref _prev) = prev_sample {
+                let time_since = (new_sample_time - prev_sample_time).as_micros() as u64;
+                let _ = tx
+                    .send(profile_event(
+                        &trace_id,
+                        &command_ref,
+                        &sample,
+                        &_prev,
+                        time_since,
+                    ))
+                    .await;
+            }
+
+            prev_sample = new_sample;
+        }
+
+        prev_sample_time = new_sample_time;
+
+        let time_to_wait = if sample_freq_ms > sampling_duration {
+            sample_freq_ms - sampling_duration
+        } else {
+            0
+        };
```

Comment:
nit: `let time_to_wait = sample_freq_ms.saturating_sub(sampling_duration)` is more concise